groups/etcd_group_service.go

```diff
@@ -0,0 +1,232 @@
+package groups
+
+import (
+	"context"
+	"encoding/json"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/coreos/etcd/clientv3/namespace"
+	"github.com/topfreegames/pitaya/config"
+	"github.com/topfreegames/pitaya/logger"
+)
+
+var (
+	clientInstance *clientv3.Client
+	leaseTTL       time.Duration
+	once           sync.Once
+	leaseID        clientv3.LeaseID
+)
+
+// EtcdGroupService base ETCD struct solution
+type EtcdGroupService struct {
+}
+
+// NewEtcdGroupService returns a new group instance
+func NewEtcdGroupService(conf *config.Config, clientOrNil *clientv3.Client) (*EtcdGroupService, error) {
+	err := initClientInstance(conf, clientOrNil)
+	if err != nil {
+		return nil, err
+	}
+	return &EtcdGroupService{}, err
+}
+
+func initClientInstance(config *config.Config, clientOrNil *clientv3.Client) error {
+	var err error
+	once.Do(func() {
+		leaseTTL = config.GetDuration("pitaya.groups.etcd.leasettl")
+		if clientOrNil != nil {
+			clientInstance = clientOrNil
+		} else {
+			clientInstance, err = createBaseClient(config)
+		}
+		if err != nil {
+			logger.Log.Fatalf("error initializing singleton etcd client in groups: %s", err.Error())
+			return
+		}
+		err = bootstrapLease()
+		if err != nil {
+			logger.Log.Fatalf("error initializing bootstrap lease in groups: %s", err.Error())
+			return
+		}
+	})
+	return err
+}
+
+func createBaseClient(config *config.Config) (*clientv3.Client, error) {
+	cli, err := clientv3.New(clientv3.Config{
+		Endpoints:   config.GetStringSlice("pitaya.groups.etcd.endpoints"),
+		DialTimeout: config.GetDuration("pitaya.groups.etcd.dialtimeout"),
+	})
+	if err != nil {
+		return nil, err
+	}
+	cli.KV = namespace.NewKV(cli.KV, config.GetString("pitaya.groups.etcd.prefix"))
+	return cli, nil
+}
+
+func bootstrapLease() error {
```

Comment:
I'm pretty sure most of this code is duplicated from the etcd service discovery. I'd rather have it all in just one place and use the common method in here and in service discovery.